File: aws/s3/typescript/src/services/tagging.ts
/**
 * Tagging Service
 *
 * S3 object and bucket tagging operations.
 */

import { S3Config, resolveEndpoint, buildPath } from "../config";
import { S3Error, mapS3ErrorCode } from "../error";
import { AwsSignerV4 } from "../signing";
import { HttpTransport, isSuccess, getHeader, getRequestId } from "../transport";
import {
  GetObjectTaggingRequest,
  PutObjectTaggingRequest,
  GetBucketTaggingRequest,
  PutBucketTaggingRequest,
  DeleteBucketTaggingRequest,
} from "../types/requests";
import {
  GetObjectTaggingOutput,
  PutObjectTaggingOutput,
  GetBucketTaggingOutput,
  PutBucketTaggingOutput,
  DeleteBucketTaggingOutput,
} from "../types/responses";
import * as xml from "../xml";
import * as crypto from "crypto";

/**
 * Tagging service for S3 object and bucket tagging operations.
 */
export class TaggingService {
  private config: S3Config;
  private transport: HttpTransport;
  private signer: AwsSignerV4;

  constructor(config: S3Config, transport: HttpTransport, signer: AwsSignerV4) {
    this.config = config;
    this.transport = transport;
    this.signer = signer;
  }

  // =============================================
  // Object Tagging Operations
  // =============================================

  /**
   * Get object tagging.
   */
  async getObjectTagging(request: GetObjectTaggingRequest): Promise<GetObjectTaggingOutput> {
    const endpoint = resolveEndpoint(this.config, request.bucket);
    const path = buildPath(this.config, request.bucket, request.key);
    const url = new URL(path, endpoint);
    url.searchParams.set("tagging", "");

    if (request.versionId) {
      url.searchParams.set("versionId", request.versionId);
    }

    const headers: Record<string, string> = {};

    const signed = await this.signer.sign("GET", url, headers);

    const response = await this.transport.send({
      method: "GET",
      url: signed.url.toString(),
      headers: signed.headers,
    });

    if (!isSuccess(response)) {
      throw this.parseError(response.body.toString());
    }

    const result = xml.parseGetObjectTagging(response.body.toString());

    return {
      versionId: getHeader(response, "x-amz-version-id"),
      tags: result.tags,
      requestId: getRequestId(response),
    };
  }

  /**
   * Put object tagging.
   */
  async putObjectTagging(request: PutObjectTaggingRequest): Promise<PutObjectTaggingOutput> {
    const endpoint = resolveEndpoint(this.config, request.bucket);
    const path = buildPath(this.config, request.bucket, request.key);
    const url = new URL(path, endpoint);
    url.searchParams.set("tagging", "");

    if (request.versionId) {
      url.searchParams.set("versionId", request.versionId);
    }

    const body = xml.buildPutTaggingXml(request.tags);
    const bodyBuffer = Buffer.from(body);
    const contentMd5 = crypto.createHash("md5").update(bodyBuffer).digest("base64");

    const headers: Record<string, string> = {
      "content-type": "application/xml",
      "content-md5": contentMd5,
      "content-length": String(bodyBuffer.length),
    };

    const signed = await this.signer.sign("PUT", url, headers, bodyBuffer);

    const response = await this.transport.send({
      method: "PUT",
      url: signed.url.toString(),
      headers: signed.headers,
      body: bodyBuffer,
    });

    if (!isSuccess(response)) {
      throw this.parseError(response.body.toString());
    }

    return {
      versionId: getHeader(response, "x-amz-version-id"),
      requestId: getRequestId(response),
    };
  }

  /**
   * Delete object tagging.
   */
  async deleteObjectTagging(bucket: string, key: string, versionId?: string): Promise<void> {
    const endpoint = resolveEndpoint(this.config, bucket);
    const path = buildPath(this.config, bucket, key);
    const url = new URL(path, endpoint);
    url.searchParams.set("tagging", "");

    if (versionId) {
      url.searchParams.set("versionId", versionId);
    }

    const headers: Record<string, string> = {};

    const signed = await this.signer.sign("DELETE", url, headers);

    const response = await this.transport.send({
      method: "DELETE",
      url: signed.url.toString(),
      headers: signed.headers,
    });

    if (!isSuccess(response)) {
      throw this.parseError(response.body.toString());
    }
  }

  // =============================================
  // Bucket Tagging Operations
  // =============================================

  /**
   * Get bucket tagging.
   */
  async getBucketTagging(request: GetBucketTaggingRequest): Promise<GetBucketTaggingOutput> {
    const endpoint = resolveEndpoint(this.config, request.bucket);
    const path = buildPath(this.config, request.bucket);
    const url = new URL(path, endpoint);
    url.searchParams.set("tagging", "");

    const headers: Record<string, string> = {};

    const signed = await this.signer.sign("GET", url, headers);

    const response = await this.transport.send({
      method: "GET",
      url: signed.url.toString(),
      headers: signed.headers,
    });

    if (!isSuccess(response)) {
      throw this.parseError(response.body.toString());
    }

    const result = xml.parseGetObjectTagging(response.body.toString()); // Same format

    return {
      tags: result.tags,
      requestId: getRequestId(response),
    };
  }

  /**
   * Put bucket tagging.
   */
  async putBucketTagging(request: PutBucketTaggingRequest): Promise<PutBucketTaggingOutput> {
    const endpoint = resolveEndpoint(this.config, request.bucket);
    const path = buildPath(this.config, request.bucket);
    const url = new URL(path, endpoint);
    url.searchParams.set("tagging", "");

    const body = xml.buildPutTaggingXml(request.tags);
    const bodyBuffer = Buffer.from(body);
    const contentMd5 = crypto.createHash("md5").update(bodyBuffer).digest("base64");

    const headers: Record<string, string> = {
      "content-type": "application/xml",
      "content-md5": contentMd5,
      "content-length": String(bodyBuffer.length),
    };

    const signed = await this.signer.sign("PUT", url, headers, bodyBuffer);

    const response = await this.transport.send({
      method: "PUT",
      url: signed.url.toString(),
      headers: signed.headers,
      body: bodyBuffer,
    });

    if (!isSuccess(response)) {
      throw this.parseError(response.body.toString());
    }

    return {
      requestId: getRequestId(response),
    };
  }

  /**
   * Delete bucket tagging.
   */
  async deleteBucketTagging(
    request: DeleteBucketTaggingRequest
  ): Promise<DeleteBucketTaggingOutput> {
    const endpoint = resolveEndpoint(this.config, request.bucket);
    const path = buildPath(this.config, request.bucket);
    const url = new URL(path, endpoint);
    url.searchParams.set("tagging", "");

    const headers: Record<string, string> = {};

    const signed = await this.signer.sign("DELETE", url, headers);

    const response = await this.transport.send({
      method: "DELETE",
      url: signed.url.toString(),
      headers: signed.headers,
    });

    if (!isSuccess(response)) {
      throw this.parseError(response.body.toString());
    }

    return {
      requestId: getRequestId(response),
    };
  }

  private parseError(body: string): S3Error {
    if (!body) {
      return new S3Error("Empty error response", "UnknownError");
    }
    const errorResponse = xml.parseErrorResponse(body);
    return mapS3ErrorCode(errorResponse.code, errorResponse);
  }
}

File: aws/s3/rust/src/services/tagging.rs
//! Tagging service for S3 object tagging operations.

use crate::config::S3Config;
use crate::error::S3Error;
use crate::signing::AwsSigner;
use crate::transport::{HttpRequest, HttpTransport};
use crate::types::*;
use crate::xml;
use bytes::Bytes;
use std::collections::HashMap;
use std::sync::Arc;
use url::Url;

/// Service for S3 object tagging operations.
pub struct TaggingService {
    config: Arc<S3Config>,
    transport: Arc<dyn HttpTransport>,
    signer: Arc<dyn AwsSigner>,
}

impl TaggingService {
    /// Create a new tagging service.
    pub fn new(
        config: Arc<S3Config>,
        transport: Arc<dyn HttpTransport>,
        signer: Arc<dyn AwsSigner>,
    ) -> Self {
        Self {
            config,
            transport,
            signer,
        }
    }

    /// Get object tagging.
    pub async fn get(
        &self,
        request: GetObjectTaggingRequest,
    ) -> Result<GetObjectTaggingOutput, S3Error> {
        let mut query = "tagging".to_string();
        if let Some(version_id) = &request.version_id {
            query = format!("{}&versionId={}", query, version_id);
        }

        let url = self.build_url(&request.bucket, Some(&request.key), Some(&query))?;
        let headers = HashMap::new();

        let signed = self.signer.sign("GET", &url, &headers, None).await?;

        let http_request = HttpRequest::new("GET", signed.url.as_str())
            .with_headers(signed.headers);

        let response = self.transport.send(http_request).await?;

        if !response.is_success() {
            return Err(self.parse_error(&response.body).await);
        }

        let body_str = String::from_utf8_lossy(&response.body);
        let mut output = xml::parse_get_object_tagging(&body_str)?;
        output.version_id = response.get_header("x-amz-version-id").map(String::from);
        output.request_id = response.request_id().map(String::from);

        Ok(output)
    }

    /// Put object tagging.
    pub async fn put(
        &self,
        request: PutObjectTaggingRequest,
    ) -> Result<PutObjectTaggingOutput, S3Error> {
        let mut query = "tagging".to_string();
        if let Some(version_id) = &request.version_id {
            query = format!("{}&versionId={}", query, version_id);
        }

        let url = self.build_url(&request.bucket, Some(&request.key), Some(&query))?;

        let body = xml::build_put_tagging_xml(&request.tags);
        let body_bytes = Bytes::from(body);
        let content_md5 = base64::encode(md5::compute(&body_bytes).0);

        let mut headers = HashMap::new();
        headers.insert("content-type".to_string(), "application/xml".to_string());
        headers.insert("content-md5".to_string(), content_md5);
        headers.insert("content-length".to_string(), body_bytes.len().to_string());

        let signed = self
            .signer
            .sign("PUT", &url, &headers, Some(&body_bytes))
            .await?;

        let http_request = HttpRequest::new("PUT", signed.url.as_str())
            .with_headers(signed.headers)
            .with_body(body_bytes);

        let response = self.transport.send(http_request).await?;

        if !response.is_success() {
            return Err(self.parse_error(&response.body).await);
        }

        Ok(PutObjectTaggingOutput {
            version_id: response.get_header("x-amz-version-id").map(String::from),
            request_id: response.request_id().map(String::from),
        })
    }

    /// Delete object tagging.
    pub async fn delete(&self, bucket: &str, key: &str, version_id: Option<&str>) -> Result<(), S3Error> {
        let mut query = "tagging".to_string();
        if let Some(vid) = version_id {
            query = format!("{}&versionId={}", query, vid);
        }

        let url = self.build_url(bucket, Some(key), Some(&query))?;
        let headers = HashMap::new();

        let signed = self.signer.sign("DELETE", &url, &headers, None).await?;

        let http_request = HttpRequest::new("DELETE", signed.url.as_str())
            .with_headers(signed.headers);

        let response = self.transport.send(http_request).await?;

        if !response.is_success() {
            return Err(self.parse_error(&response.body).await);
        }

        Ok(())
    }

    fn build_url(
        &self,
        bucket: &str,
        key: Option<&str>,
        query: Option<&str>,
    ) -> Result<Url, S3Error> {
        let endpoint = self.config.resolve_endpoint(Some(bucket));
        let path = self.config.build_path(bucket, key);

        let url_str = if let Some(q) = query {
            format!("{}{}?{}", endpoint.as_str().trim_end_matches('/'), path, q)
        } else {
            format!("{}{}", endpoint.as_str().trim_end_matches('/'), path)
        };

        Url::parse(&url_str).map_err(|e| {
            S3Error::Request(crate::error::RequestError::Validation {
                message: format!("Invalid URL: {}", e),
            })
        })
    }

    async fn parse_error(&self, body: &Bytes) -> S3Error {
        if body.is_empty() {
            return S3Error::Response(crate::error::ResponseError::InvalidResponse {
                message: "Empty error response".to_string(),
            });
        }

        let body_str = String::from_utf8_lossy(body);
        match xml::parse_error_response(&body_str) {
            Ok(error_response) => {
                crate::error::map_s3_error_code(&error_response.code, Some(error_response))
            }
            Err(_) => S3Error::Response(crate::error::ResponseError::InvalidResponse {
                message: format!(
                    "Failed to parse error response: {}",
                    body_str.chars().take(100).collect::<String>()
                ),
            }),
        }
    }
}

impl std::fmt::Debug for TaggingService {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("TaggingService")
            .field("config", &self.config)
            .finish_non_exhaustive()
    }
}
